### PR TITLE
Feature/hbase namespaces integration queues

### PR DESCRIPTION
**Pull request by anwar6953**
Use HBase namespaces for queues/stream tables.

JIRA: https://issues.cask.co/browse/CDAP-1594
Build: http://builds.cask.co/browse/CDAP-DUT908-26

### CODE REVIEW

File: cdap-hbase-compat-0.94/src/main/java/co/cask/cdap/data2/util/hbase/HBase94TableUtil.java [hunk 1]
```diff
@@ -125,6 +125,26 @@ public void modifyTable(HBaseAdmin admin, HTableDescriptor tableDescriptor) thro
   }
 
   @Override
+  public void deleteAllInNamespace(HBaseAdmin admin, Id.Namespace namespaceId, String tablePrefix) throws IOException {
+    // Note: the following method call is incorrect
```
Comment (bdmogal) [inline cdap-hbase-compat-0.94/src/main/java/co/cask/cdap/data2/util/hbase/HBase94TableUtil.java]: Remove

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 2]
```diff
@@ -244,17 +244,33 @@ public void upgrade(String tableName, Properties properties) throws Exception {
...
 
+  public TableId getConfigTableId(QueueName queueName) {
+    return getConfigTableId(queueName.getFirstComponent());
+  }
+
+  public TableId getConfigTableId(String namespace) {
+    return TableId.from(getConfigTableName(namespace));
+  }
+
+  public TableId getDataTableId(QueueName queueName) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: any way to unify this and the method below? Use overloading so that their names are the same (and - hopefully - one calls the other?)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamUtils.java [hunk 3]
```diff
@@ -430,7 +430,8 @@ public Long getResult() {
    * @return constructed table name
    */
   public static String getStateStoreTableName(Id.Namespace namespace) {
-    return String.format("cdap.%s.%s.state.store", namespace.getId(), QueueConstants.QueueType.STREAM.toString());
+    return String.format("cdap.%s.system.%s.state.store",
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamUtils.java]: Is it right? I thought you need to go through the `HTableNameConverter`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java [hunk 4]
```diff
@@ -45,19 +46,34 @@ public String getTablePrefix() {
     return namespace;
   }
 
+  @VisibleForTesting
+  public String getHBaseNamespace() {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java]: I would rather not have this method. For testing, you can simply call `HBaseTableUtil.toBaseNamespace(id.getNamespace())`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java [hunk 5]
```diff
@@ -27,7 +27,7 @@
  * Common implementation of table-based QueueAdmin
  */
 public abstract class AbstractQueueAdmin implements QueueAdmin {
-  private final String unqualifiedTableNamePrefix;
+  protected final String unqualifiedTableNamePrefix;
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java]: Why need to keep this? I thought we want to use `TableId` everywhere?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/QueueUtils.java [hunk 6]
```diff
@@ -24,12 +24,22 @@ public static String determineQueueConfigTableName(String queueTableName) {
     // the name of this table has the form: <cdap root namespace>.<queue namespace>.<system namespace>.(queue|stream).*
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/QueueUtils.java]: You might want to update the comment to reflect the code change.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 7]
```diff
@@ -192,21 +193,20 @@ public void truncate(String name) throws Exception {
...
-  private void truncate(byte[] tableNameBytes) throws IOException {
+  private void truncate(TableId tableId) throws IOException {
     HBaseAdmin admin = getHBaseAdmin();
-    if (admin.tableExists(tableNameBytes)) {
-      HTableDescriptor tableDescriptor = admin.getTableDescriptor(tableNameBytes);
-      admin.disableTable(tableNameBytes);
-      admin.deleteTable(tableNameBytes);
-      admin.createTable(tableDescriptor);
+    if (tableUtil.tableExists(admin, tableId)) {
+      HTableDescriptor tableDescriptor = tableUtil.getHTableDescriptor(admin, tableId);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: The method should be called `createHTableDescriptor`.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: oh, you actually getting, not creating new one. That's ok. Let's keep the name.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: I looked further, you are creating... so, rename it to `createHTableDescriptor`. Sorry for the confusion.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: I got confused by the naming since you have to method named the same, but actually one is create, one is get. You should name them as:

```java
getHTableDescriptor(HBaseAdmin, TableId tableId);   // Get existing table descriptor
createHTableDescriptor(TableId tableId);                     // Create new instance of HTableDescriptor for the given tableId
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 8]
```diff
@@ -192,21 +193,20 @@ public void truncate(String name) throws Exception {
...
+      tableUtil.disableTable(admin, tableId);
+      tableUtil.deleteTable(admin, tableId);
+      tableUtil.createTableIfNotExists(admin, tableId, tableDescriptor);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: It's better to have a `truncate(HBaseAdmin, TableId)` method on `HBaseTableUtil`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 9]
```diff
@@ -236,7 +234,9 @@ public void drop(String name) throws Exception {
 
   @Override
   public void upgrade(String tableName, Properties properties) throws Exception {
-    AbstractHBaseDataSetAdmin dsAdmin = new DatasetAdmin(tableName, hConf, tableUtil, properties);
+    //TODO: check this
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: What is this TODO about? If it is truly a TODO, there should be a JIRA associated with it.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 10]
```diff
@@ -236,7 +234,9 @@ public void drop(String name) throws Exception {
 
   @Override
   public void upgrade(String tableName, Properties properties) throws Exception {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Upgrade should not takes table name. It should just take a namespace and the upgrade should performs upgrade all queue tables under that namespace.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Otherwise the one who calls upgrade needs to know what table is a queue table, which is wrong. The queue table(s) should be encapsulated in this class, not exposed.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 11]
```diff
@@ -244,17 +244,33 @@ public void upgrade(String tableName, Properties properties) throws Exception {
...
 
-  private void drop(byte[] tableName) throws IOException {
+  private void drop(TableId tableId) throws IOException {
     HBaseAdmin admin = getHBaseAdmin();
-    if (admin.tableExists(tableName)) {
-      admin.disableTable(tableName);
-      admin.deleteTable(tableName);
+    if (tableUtil.tableExists(admin, tableId)) {
+      tableUtil.disableTable(admin, tableId);
+      tableUtil.deleteTable(admin, tableId);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Better move it to `HBaseTableUtil.drop(TableId)`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 12]
```diff
@@ -244,17 +244,33 @@ public void upgrade(String tableName, Properties properties) throws Exception {
...
+  }
+
+  public TableId getTableIdFromFlow(String namespaceId, String app, String flow) {
+    return TableId.from(getTableNameForFlow(namespaceId, app, flow));
+  }
+
   private void deleteConsumerConfigurations(QueueName queueName) throws IOException {
     // we need to delete the row for this queue name from the config table
-    HTable hTable = new HTable(getHBaseAdmin().getConfiguration(), getConfigTableName(queueName));
+    HTable hTable = tableUtil.getHTable(getHBaseAdmin().getConfiguration(), getConfigTableId(queueName));
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Should be `createHTable`, not get.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: In general, if you are replacing a "new XXX" to a method on util class, it should be "createXXX".

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 13]
```diff
@@ -277,9 +293,10 @@ private void deleteConsumerConfigurationsForPrefix(String tableNamePrefix, Strin
     throws IOException {
     // table is created lazily, possible it may not exist yet.
     HBaseAdmin admin = getHBaseAdmin();
-    if (admin.tableExists(configTableName)) {
+    TableId tableId = TableId.from(configTableName);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Shouldn't create TableId by parsing a name.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java [hunk 14]
```diff
@@ -367,8 +384,10 @@ private void createConfigTable(String configTableName) throws IOException {
   @Override
   public void dropAllInNamespace(String namespaceId) throws Exception {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Should modify the interface to take `Id.Namespace` rather than a string.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java [hunk 15]
```diff
@@ -109,7 +114,7 @@ private HBaseQueueAdmin ensureTableExists(QueueName queueName) throws IOExceptio
   }
 
   private HTable createHTable(String name) throws IOException {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java]: Should get rid of this method and call `HBaseQueueAdmin` to do it (by taking a `QueueName` rather than a `String`).

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java [hunk 16]
```diff
@@ -31,6 +31,7 @@
 import com.google.common.hash.Hashing;
 import com.google.common.io.Files;
 import com.google.common.io.OutputSupplier;
+import joptsimple.internal.Strings;
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java]: What is it?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java [hunk 17]
```diff
@@ -62,6 +63,7 @@
 import java.util.jar.JarEntry;
 import java.util.jar.JarOutputStream;
 import java.util.regex.Matcher;
+import java.util.regex.Pattern;
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java]: I don't see it's being used.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java [hunk 18]
```diff
@@ -45,19 +46,34 @@ public String getTablePrefix() {
...
+    return HBaseTableUtil.toHBaseNamespace(namespace);
+  }
+
   /**
-   *@return Backward compatible, ASCII encoded table name
+   * @return Backward compatible, ASCII encoded table name
    */
-  public String getTableName() {
+  public String getHBaseTableName() {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java]: TableId shouldn't have anything to do with HBase.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java [hunk 19]
```diff
@@ -76,6 +92,45 @@ public static TableId from(String tablePrefix, String namespace, String tableNam
     return new TableId(tablePrefix, Id.Namespace.from(namespace), tableName);
   }
 
+  public static TableId from(String hBaseNamespace, String hTableName) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java]: This logic should be in this class. They should go into corresponding `HTableNameConverter`

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java [hunk 20]
```diff
@@ -332,7 +334,10 @@ public void testPrefix() {
 
   @Override
   protected void forceEviction(QueueName queueName) throws Exception {
-    byte[] tableName = Bytes.toBytes(((HBaseQueueClientFactory) queueClientFactory).getTableName(queueName));
+    // This functionality (internal-to-cdap tableName -> hbase tableName conversion) should be localized
+    // It's also used in HBaseQueueTest#getConsumerConfigCache
+    String tableNameString = ((HBaseQueueClientFactory) queueClientFactory).getTableName(queueName);
```
Comment (chtyim) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java]: Should change `getTableName` to `getTableId`. We should get rid of the `TableId.from(String)` method.

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java [hunk 21]
```diff
@@ -317,8 +305,22 @@ protected void verifyConsumerConfigIsDeleted(QueueName... queueNames) throws Exc
     }
   }
 
+  private ConsumerConfigCache getConsumerConfigCache(QueueName queueName) throws Exception {
+    TableId tableId = TableId.from(((HBaseQueueAdmin) queueAdmin).getConfigTableName(queueName));
```
Comment (chtyim) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java]: Make `HBaseQueueAdmin.getConfigTableName` to `HBaseQueueAdmin.getConfigTableId`.

File: cdap-hbase-compat-0.94/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase94/HBaseQueueRegionObserver.java [hunk 22]
```diff
@@ -79,7 +80,8 @@ public void start(CoprocessorEnvironment env) {
         this.prefixBytes = HBaseQueueAdmin.SALT_BYTES;
       }
 
-      namespaceId = HBaseQueueAdmin.getNamespaceId(tableName);
+      namespaceId = HBase94TableUtil.fromTableName(tableDesc.getNameAsString()).getNamespace().getId();
```
Comment (chtyim) [inline cdap-hbase-compat-0.94/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase94/HBaseQueueRegionObserver.java]: RegionObserver shouldn't have dependency on table util. Move that method to `HTableNameConverter`, which is what it is intended for.

**Comment by chtyim** [comment]
Done first round of review. Please address the comment.

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java (hunk 20 above)
Comment (bdmogal) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java]: @chtyim I plan to remove the ``TableId.from(String)`` method after the changing dataset APIs to have namespace. 

@anwar6953 Please do not use that method in this PR. I'll get rid of it once the dataset API changes are in.

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java (hunk 21 above)
Comment (anwar6953) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java]: done

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java (hunk 20 above)
Comment (anwar6953) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueTest.java]: done. Still need to get rid of the TableId.from(String) though.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java (hunk 17 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java]: removed (was in a comment that is now outdated).

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java (hunk 16 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HBaseTableUtil.java]: removed. using inline `"`" now.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 14 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: this change is a little larger, so will do as separate PR: https://issues.cask.co/browse/CDAP-1483

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/QueueUtils.java (hunk 6 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/QueueUtils.java]: Done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 7 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: done

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 9 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: This was just an artifact of my work-in-progress.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 10 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: There's an upgrade() method in this class that does this.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 12 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 10 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Then this method shouldn't be public?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java (hunk 19 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java]: moved.

File: cdap-hbase-compat-0.94/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase94/HBaseQueueRegionObserver.java (hunk 22 above)
Comment (anwar6953) [inline cdap-hbase-compat-0.94/src/main/java/co/cask/cdap/data2/transaction/queue/coprocessor/hbase94/HBaseQueueRegionObserver.java]: moved.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 10 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: Ok. I'm not sure why it was public on the interface.
Maybe the designer of interface wanted to allow also upgrading of specific tables, instead of all (if all is too many).

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: I  think it was inherited from DatasetAdmin. Since now the QueueAdmin no longer extends from DatasetAdmin, I think can be changed to private.

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamUtils.java (hunk 3 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamUtils.java]: In the base, I left it like this, so levelDB, inmemory can still use this.
And then only HBase version uses TableId.

Alternative would be to have a method that turns TableId into a full tableName (such as in hbase 94).

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java (hunk 4 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java]: removed. (usages now use HTableNameConverter.toHBaseNamespace).

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java (hunk 18 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/TableId.java]: Moved to HTableNameConverter.getHBaseTableName(

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java (hunk 5 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java]: I left the base class to use String, for levelDb/inmemory, and only HBase uses TableId.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java (hunk 8 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueAdmin.java]: done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HTableNameConverter.java [hunk 23]
```diff
@@ -16,18 +16,87 @@
...
+import com.google.common.base.Preconditions;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+
 /**
  * Common utility methods for dealing with HBase table name conversions.
  */
 public abstract class HTableNameConverter {
+  protected static final String HBASE_NAMESPACE_PREFIX = "cdap_";
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HTableNameConverter.java]: I thought we always use the `root.namespace` from cConf. Is that changed?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java [hunk 24]
```diff
@@ -54,9 +55,9 @@
   private final HTable hTable;
 
   public HBaseMetricsTable(String name, Configuration hConf) throws IOException {
-    String hTableName = HBaseTableUtil.getHBaseTableName(name);
+    String hTableName = HTableNameConverter.getHBaseTableName(name);
     HBaseTableUtil tableUtil = new HBaseTableUtilFactory().get();
-    HTable hTable = tableUtil.getHTable(hConf, TableId.from(hTableName));
+    HTable hTable = tableUtil.createHTable(hConf, TableId.from(hTableName));
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java]: Shouldn't use `TableId.from`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java [hunk 25]
```diff
@@ -79,43 +60,32 @@ public static String getFlowName(String queueTableName) {
     return parts[parts.length - 1];
   }
 
+  public TableId getConfigTableId(QueueName queueName) {
+    return getConfigTableId(queueName.getFirstComponent());
+  }
+
+  public TableId getConfigTableId(String namespace) {
+    return TableId.from(root, namespace, unqualifiedConfigTableNameSuffix);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java]: TableId should be aware of "root" prefix. It should be a simple POJO that has namespace and name only.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java (hunk 24 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java]: Yes, this will be fixed when we have new ``DatasetDefinition`` APIs available. Coming in another PR.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HTableNameConverter.java (hunk 23 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/util/hbase/HTableNameConverter.java]: There's a JIRA for that, separate from the subject of this PR: https://issues.cask.co/browse/CDAP-1670

So it will get done in a separate PR by either Bhooshan or me.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java (hunk 25 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java]: How would we attain this? Inject the CConf into HBaseTableUtil somehow?

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java]: Also, since its not on the same topic (And to keep this PR small and merged without taking too long), I prefer to do this in a separate PR.

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/AbstractQueueAdmin.java]: https://issues.cask.co/browse/CDAP-1688

**Comment by anwar6953** [comment]
Closing. Changes presented in this PR are split up into two separate PRs:
https://github.com/caskdata/cdap/pull/1491
